controllers/topology/desired_state.go

```diff
@@ -177,14 +192,77 @@ func computeControlPlane(_ context.Context, s *scope.Scope, infrastructureMachin
 	}
 
 	// Sets the desired Kubernetes version for the control plane.
-	// TODO: improve this logic by adding support for version upgrade component by component
-	if err := contract.ControlPlane().Version().Set(controlPlane, s.Blueprint.Topology.Version); err != nil {
+	version, err := computeControlPlaneVersion(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compute version of control plane")
+	}
+	if err := contract.ControlPlane().Version().Set(controlPlane, version); err != nil {
 		return nil, errors.Wrap(err, "failed to set spec.version in the ControlPlane object")
 	}
 
 	return controlPlane, nil
 }
 
+// computeControlPlaneVersion calculates the version of the desired control plane.
+// The version if calculated using the the state of the current machine deployments, the current control plane
+// and the version defined in the topology.
+func computeControlPlaneVersion(s *scope.Scope) (string, error) {
+	desiredVersion := s.Blueprint.Topology.Version
+	// creating a new control plane object, use version from topology
+	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
+		return desiredVersion, nil
+	}
+
+	// get the current version of the control plane
+	version, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get the version from control plane spec")
+	}
+
+	// return early if the version is already equal to the desired version
+	// no further checks required.
+	if *version == desiredVersion {
+		return *version, nil
+	}
+
+	// check if the current control plane is upgrading
+	cpUpgrading, err := contract.ControlPlane().IsUpgrading(s.Current.ControlPlane.Object)
+	if err != nil {
+		return *version, errors.Wrap(err, "failed to check if control plane is upgrading")
+	}
+	// If the current control plane is upgrading, then do not pick up the desiredVersion yet.
+	// Return the current version of the control plane. We will pickup the new version after the control
+	// plane is stable.
+	if cpUpgrading {
+		return *version, nil
+	}
+
+	// Control plane supports replicas. Check if the control plane is in the middle of a scale operation.
+	if s.Blueprint.Topology.ControlPlane.Replicas != nil {
+		cpScaling, err := contract.ControlPlane().IsScaling(s.Current.ControlPlane.Object)
+		if err != nil {
+			return *version, errors.Wrap(err, "failed to check if the control plane is scaling")
```

Comment:
```suggestion
			return "", errors.Wrap(err, "failed to check if the control plane is scaling")
```
Nit (up to you) I think it would be better to return an empty string whenever we also return an error (also affects other returns below)